You have to work on the [University dataset](https://drive.google.com/drive/folders/1Hs3nRtK_F3h8eg59B4-TD1DEua6g8Klv?usp=sharing). It contains three different university rankings:
*  The Times Higher Education World University Ranking, shortly *Times*,
*  the Academic Ranking of World Universities, shortly *Shanghai*,
*  the Center for World University Rankings, shortly *cwur*.

### Notes

1.    It is mandatory to use GitHub for developing the project.
1.    The project must be a jupyter notebook.
1.    There is no restriction on the libraries that can be used, nor on the Python version.
1.    All questions on the project **must** be asked in a public channel on [Zulip](https://focs.zulipchat.com).

In [1]:
import pandas as pd

Caricamento del file *cwurData.csv* da Google Drive, rinominando la colonna dei nomi delle università da `institution` a `university_name` in modo da avere lo stesso nome rispetto agli altri file dei ranking.

In [2]:
cwur = pd.read_csv('cwurData.csv')
cwur.rename(columns = {'institution':'university_name'}, inplace = True)
cwur.head()

world_rank                        university_name         country  \
0           1                     Harvard University             USA   
1           2  Massachusetts Institute of Technology             USA   
2           3                    Stanford University             USA   
3           4                University of Cambridge  United Kingdom   
4           5     California Institute of Technology             USA   

   national_rank  quality_of_education  alumni_employment  quality_of_faculty  \
0              1                     7                  9                   1   
1              2                     9                 17                   3   
2              3                    17                 11                   5   
3              1                    10                 24                   4   
4              4                     2                 29                   7   

   publications  influence  citations  broad_impact  patents   score  year  
0             1          1          1           NaN        5  100.00  2012  
1            12          4          4           NaN        1   91.67  2012  
2             4          2          2           NaN       15   89.50  2012  
3            16         16         11           NaN       50   86.17  2012  
4            37         22         22           NaN       18   85.21  2012

Caricamento dell file *education_expenditure_supplementary_data.csv* dalla cartella di Google Drive condivisa, impostando "\r" come carattere di fine riga per una lettura corretta.

In [3]:
expenditure = pd.read_csv('education_expenditure_supplementary_data.csv', lineterminator = '\r')
expenditure.head()

country     institute_type direct_expenditure_type  1995  2000  2005  \
0  OECD Average  All Institutions                   Public   4.9   4.9   5.0   
1     Australia  All Institutions                   Public   4.5   4.6   4.3   
2       Austria  All Institutions                   Public   5.3   5.4   5.2   
3       Belgium  All Institutions                   Public   5.0   5.1   5.8   
4        Canada  All Institutions                   Public   5.8   5.2   4.8   

   2009  2010  2011  
0   5.4   5.4   5.3  
1   4.5   4.6   4.3  
2   5.7   5.6   5.5  
3   6.4   6.4   6.4  
4   5.0   5.2   NaN

Caricamento del file *educational_attainment_supplementary_data.csv* dalla cartella di Google Drive condivisa.

In [4]:
attainment = pd.read_csv('educational_attainment_supplementary_data.csv')
attainment.head()

country_name                                        series_name  1985  1986  \
0  Afghanistan  Barro-Lee: Average years of primary schooling,...  0.33   NaN   
1  Afghanistan  Barro-Lee: Average years of primary schooling,...  1.03   NaN   
2  Afghanistan  Barro-Lee: Average years of primary schooling,...  0.83   NaN   
3  Afghanistan  Barro-Lee: Average years of primary schooling,...  2.34   NaN   
4  Afghanistan  Barro-Lee: Average years of primary schooling,...  0.54   NaN   

   1987  1990  1991  1992  1993  1995  ...  2005  2006  2007  2008  2009  \
0   NaN  0.44   NaN   NaN   NaN  0.57  ...  0.86   NaN   NaN   NaN   NaN   
1   NaN  1.26   NaN   NaN   NaN  1.54  ...  2.18   NaN   NaN   NaN   NaN   
2   NaN  0.95   NaN   NaN   NaN  1.26  ...  1.01   NaN   NaN   NaN   NaN   
3   NaN  2.22   NaN   NaN   NaN  2.37  ...  2.26   NaN   NaN   NaN   NaN   
4   NaN  0.92   NaN   NaN   NaN  0.94  ...  2.00   NaN   NaN   NaN   NaN   

   2010  2011  2012  2013  2015  
0  1.27   NaN   NaN   NaN   NaN  
1  2.64   NaN   NaN   NaN   NaN  
2  2.45   NaN   NaN   NaN   NaN  
3  3.55   NaN   NaN   NaN   NaN  
4  1.29   NaN   NaN   NaN   NaN  

[5 rows x 29 columns]

Caricamento del file *school_and_country_table.csv* dalla cartella di Google Drive condivisa.

In [5]:
schoolCountry = pd.read_csv('school_and_country_table.csv')
print(schoolCountry.shape)
schoolCountry.head()

(818, 2)


school_name                   country
0                     Harvard University  United States of America
1     California Institute of Technology  United States of America
2  Massachusetts Institute of Technology  United States of America
3                    Stanford University  United States of America
4                   Princeton University  United States of America

Caricamento del file *shanghaiData.csv* dalla cartella di Google Drive condivisa.

In [6]:
shanghai = pd.read_csv('shanghaiData.csv')
shanghai.head()

world_rank                              university_name national_rank  \
0          1                           Harvard University             1   
1          2                      University of Cambridge             1   
2          3                          Stanford University             2   
3          4           University of California, Berkeley             3   
4          5  Massachusetts Institute of Technology (MIT)             4   

   total_score  alumni  award   hici     ns    pub   pcp  year  
0        100.0   100.0  100.0  100.0  100.0  100.0  72.4  2005  
1         73.6    99.8   93.4   53.3   56.6   70.9  66.9  2005  
2         73.4    41.1   72.2   88.5   70.9   72.3  65.0  2005  
3         72.8    71.8   76.0   69.4   73.9   72.2  52.7  2005  
4         70.1    74.0   80.6   66.7   65.8   64.3  53.0  2005

Caricamento del file *timesData.csv* dalla cartella di Google Drive condivisa: la colonna `num_students` presenta la virgola come separatore delle migliaia, mentre la colonna `female_male_ratio` presenta in alcune righe solamente il carattere "-" che quindi sarà aggiunto alla sequenza di caratteri interpretati come valori mancanti.

In [7]:
times = pd.read_csv('timesData.csv', thousands = ',', na_values = '-')
times.head()

world_rank                        university_name                   country  \
0          1                     Harvard University  United States of America   
1          2     California Institute of Technology  United States of America   
2          3  Massachusetts Institute of Technology  United States of America   
3          4                    Stanford University  United States of America   
4          5                   Princeton University  United States of America   

   teaching  international  research  citations  income  total_score  \
0      99.7           72.4      98.7       98.8    34.5         96.1   
1      97.7           54.6      98.0       99.9    83.7         96.0   
2      97.8           82.3      91.4       99.9    87.5         95.6   
3      98.3           29.5      98.1       99.2    64.3         94.3   
4      90.9           70.3      95.4       99.9     NaN         94.2   

   num_students  student_staff_ratio international_students female_male_ratio  \
0       20152.0                  8.9                    25%               NaN   
1        2243.0                  6.9                    27%           33 : 67   
2       11074.0                  9.0                    33%           37 : 63   
3       15596.0                  7.8                    22%           42 : 58   
4        7929.0                  8.4                    27%           45 : 55   

   year  
0  2011  
1  2011  
2  2011  
3  2011  
4  2011

## Preprocessing

Prima di effettuare qualsiasi tipo di operazione sono stati effettuati dei controlli per verificare la correttezza dei nomi delle università. <br>
Uno dei problemi rilevati è stata la punteggiatura: infatti anche all'interno della stessa classifica un singolo carattere può fare la differenza tra due nomi che indicano la stessa università.

In [8]:
shanghai.iloc[[3115, 3614, 3997, 4497]]

world_rank           university_name national_rank  total_score  alumni  \
3115    102-150  Aix-Marseille University           4-6          NaN    15.6   
3614    101-150  Aix Marseille University           4-7          NaN    14.3   
3997    101-150  Aix Marseille University           5-7          NaN    13.9   
4497    101-150  Aix Marseille University           5-6          NaN    13.6   

      award  hici    ns   pub   pcp  year  
3115    0.0  16.1  22.4  44.3  15.3  2011  
3614    0.0  16.2  22.1  44.3  15.3  2012  
3997    0.0  16.8  19.5  50.0  26.5  2014  
4497    0.0  15.2  20.3  50.1  27.6  2015

Un po' più frequente è il caso in cui la stessa università viene riportata utilizzando una punteggiatura differente in due classifiche diverse.

In [9]:
print(times.loc[1357]['university_name'])
print(cwur.loc[997]['university_name'])

China Medical University, Taiwan
China Medical University (Taiwan)


Le maiuscole/minuscole creano delle differenze tra due stringhe, soprattutto se utilizzati come chiave di unione tra i dataset.

In [10]:
print(times.loc[836]['university_name'])
print(shanghai.loc[4326]['university_name'])

Mines ParisTech
MINES ParisTech


Quindi si è optato per la creazione di una nuova colonna con i nomi delle università, la quale viene creata secondo i seguenti criteri:
1. sostituzione di tutti i caratteri di punteggiatura con degli spazi
2. se nel nel punto 1 è stata creata una sequenza di spazi, essa viene ridefinita allo spazio singolo
3. se nel nel punto 1 sono stati creati degli spazi all'inizio o alla fine della stringa, questi vengono eliminati
4. tutti i caratteri sono trasformati in minuscolo

In [11]:
times['univ_name'] = times['university_name'].str.replace('[^\w\s]',' ').str.replace('_',' ').str.replace(' +',' ').str.strip().str.lower()
shanghai['univ_name'] = shanghai['university_name'].str.replace('[^\w\s]',' ').str.replace('_',' ').str.replace(' +',' ').str.strip().str.lower()
cwur['univ_name'] = cwur['university_name'].str.replace('[^\w\s]',' ').str.replace('_',' ').str.replace(' +',' ').str.strip().str.lower()

In [12]:
times[['university_name', 'univ_name']].head()

university_name  \
0                     Harvard University   
1     California Institute of Technology   
2  Massachusetts Institute of Technology   
3                    Stanford University   
4                   Princeton University   

                               univ_name  
0                     harvard university  
1     california institute of technology  
2  massachusetts institute of technology  
3                    stanford university  
4                   princeton university

Sono stati notati inoltre degli errori di digitazione per alcuni valori dell'attributo `country` della classifica *Times* (utilizzato nei punti successivi).

In [13]:
times.iloc[[2019, 115, 2205, 551]]

world_rank                 university_name                    country  \
2019    201-250  University of Hawai’i at Mānoa  Unisted States of America   
115         115  University of Hawai’i at Mānoa   United States of America   
2205    401-500                Aston University              Unted Kingdom   
551     350-400                Aston University             United Kingdom   

      teaching  international  research  citations  income  total_score  \
2019      32.0           63.4      35.5       67.7    44.7          NaN   
115       38.3           34.2      47.6       81.0     NaN         55.2   
2205      18.4           92.5      20.4       51.9    34.6          NaN   
551       16.2           84.5      19.9       11.7    26.8          NaN   

      num_students  student_staff_ratio international_students  \
2019           NaN                  NaN                    NaN   
115            NaN                  NaN                    NaN   
2205           NaN                  NaN                    NaN   
551            NaN                  NaN                    NaN   

     female_male_ratio  year                       univ_name  
2019               NaN  2016  university of hawai i at mānoa  
115                NaN  2011  university of hawai i at mānoa  
2205               NaN  2016                aston university  
551                NaN  2012                aston university

In [14]:
times['country'] = times['country'].str.replace('Unisted', 'United')
times['country'] = times['country'].str.replace('Unted', 'United')

In [15]:
times.iloc[[2019, 115, 2205, 551]]

world_rank                 university_name                   country  \
2019    201-250  University of Hawai’i at Mānoa  United States of America   
115         115  University of Hawai’i at Mānoa  United States of America   
2205    401-500                Aston University            United Kingdom   
551     350-400                Aston University            United Kingdom   

      teaching  international  research  citations  income  total_score  \
2019      32.0           63.4      35.5       67.7    44.7          NaN   
115       38.3           34.2      47.6       81.0     NaN         55.2   
2205      18.4           92.5      20.4       51.9    34.6          NaN   
551       16.2           84.5      19.9       11.7    26.8          NaN   

      num_students  student_staff_ratio international_students  \
2019           NaN                  NaN                    NaN   
115            NaN                  NaN                    NaN   
2205           NaN                  NaN                    NaN   
551            NaN                  NaN                    NaN   

     female_male_ratio  year                       univ_name  
2019               NaN  2016  university of hawai i at mānoa  
115                NaN  2011  university of hawai i at mānoa  
2205               NaN  2016                aston university  
551                NaN  2012                aston university

# 1. For each university, extract from the *times* dataset the most recent and the least recent data, obtaining two separate dataframes

Per estrarre i dati più vecchi, per ogni università viene selezionato l'anno più piccolo.

In [16]:
old_times = times.iloc[times.groupby('univ_name')['year'].idxmin()]
old_times.head()

world_rank             university_name         country  teaching  \
501     301-350          Aalborg University         Denmark      19.0   
502     301-350            Aalto University         Finland      26.2   
166         167           Aarhus University         Denmark      38.1   
476     276-300      Aberystwyth University  United Kingdom      19.8   
2404    601-800  Adam Mickiewicz University          Poland      20.0   

      international  research  citations  income  total_score  num_students  \
501            75.3      20.0       27.1    36.4          NaN       17422.0   
502            49.0      22.2       37.5    61.9          NaN       16099.0   
166            33.4      55.6       57.3    61.5         49.9       23895.0   
476            63.8      15.5       56.6    35.5          NaN        9252.0   
2404           25.7      11.0       15.3    28.7          NaN       40633.0   

      student_staff_ratio international_students female_male_ratio  year  \
501                  15.9                    15%           48 : 52  2012   
502                  24.2                    17%           32 : 68  2012   
166                  13.6                    14%           54 : 46  2011   
476                  19.2                    18%           48 : 52  2012   
2404                 15.6                     1%           71 : 29  2016   

                       univ_name  
501           aalborg university  
502             aalto university  
166            aarhus university  
476       aberystwyth university  
2404  adam mickiewicz university

Per estrarre i dati più recenti, per ogni università viene selezionato l'anno più grande.

In [17]:
recent_times = times.iloc[times.groupby('univ_name')['year'].idxmax()].copy()
recent_times.head()

world_rank             university_name         country  teaching  \
2003    201-250          Aalborg University         Denmark      25.1   
2056    251-300            Aalto University         Finland      31.1   
1908       =106           Aarhus University         Denmark      36.9   
2105    301-350      Aberystwyth University  United Kingdom      21.6   
2404    601-800  Adam Mickiewicz University          Poland      20.0   

      international  research  citations  income  total_score  num_students  \
2003           71.0      28.4       73.8    43.7          NaN       17422.0   
2056           65.4      32.8       62.1    61.6          NaN       16099.0   
1908           76.8      50.7       79.8    68.3         57.7       23895.0   
2105           72.2      18.9       67.2    31.3          NaN        9252.0   
2404           25.7      11.0       15.3    28.7          NaN       40633.0   

      student_staff_ratio international_students female_male_ratio  year  \
2003                 15.9                    15%           48 : 52  2016   
2056                 24.2                    17%           32 : 68  2016   
1908                 13.6                    14%           54 : 46  2016   
2105                 19.2                    18%           48 : 52  2016   
2404                 15.6                     1%           71 : 29  2016   

                       univ_name  
2003          aalborg university  
2056            aalto university  
1908           aarhus university  
2105      aberystwyth university  
2404  adam mickiewicz university

> Si noti che non tutte le università sono entrate nel ranking nello stesso anno, quindi i dati più vecchi non fanno riferimento ad un anno comune. Lo stesso vale per i dati più recenti.

# 2. For each university, compute the improvement in `income` between the least recent and the most recent data points

Per ottenere il miglioramento di introiti per ciascuna università, vengono uniti i dataset del punto precedente utilizzando `univ_name` come chiave, poi viene sottratto l'introito dell'anno più vecchio al dato più recente.

In [18]:
fusion_income = pd.merge(old_times, recent_times, on = ['univ_name', 'country'], suffixes = ('_old', '_recent'))
fusion_income['improvement'] = fusion_income['income_recent'] - fusion_income['income_old']
fusion_income[['university_name_old', 'country', 'year_old', 'income_old', 'year_recent', 'income_recent', 'improvement']]

university_name_old         country  year_old  \
0                          Aalborg University         Denmark      2012   
1                            Aalto University         Finland      2012   
2                           Aarhus University         Denmark      2011   
3                      Aberystwyth University  United Kingdom      2012   
4                  Adam Mickiewicz University          Poland      2016   
..                                        ...             ...       ...   
813                  École Normale Supérieure          France      2011   
814          École Normale Supérieure de Lyon          France      2011   
815                       École Polytechnique          France      2011   
816  École Polytechnique Fédérale de Lausanne     Switzerland      2011   
817                         Örebro University          Sweden      2016   

     income_old  year_recent  income_recent  improvement  
0          36.4         2016           43.7          7.3  
1          61.9         2016           61.6         -0.3  
2          61.5         2016           68.3          6.8  
3          35.5         2016           31.3         -4.2  
4          28.7         2016           28.7          0.0  
..          ...          ...            ...          ...  
813        30.7         2016           37.1          6.4  
814        26.1         2016           31.7          5.6  
815         NaN         2016           82.3          NaN  
816        38.0         2016           65.4         27.4  
817        29.8         2016           29.8          0.0  

[818 rows x 7 columns]

> Le università con `improvement` positivo hanno avuto un miglioramento in quantità di introiti rispetto ai dati più vecchi, quelle con `improvement` negativo hanno registrato un calo delle entrate. <br>
Il valore è nullo se non è stato possibile calcolare la differenza a causa di valori mancanti.

# 3. Find the university with the largest increase computed in the previous point

Per ottenere l'incremento più alto registrato si estraggono le università che hanno `improvement` pari al valore maggiore.

In [19]:
fusion_income.loc[fusion_income['improvement'] >= fusion_income['improvement'].max()][['university_name_old', 'country', 'year_old', 'income_old', 'year_recent', 'income_recent', 'improvement']]

university_name_old  country  year_old  income_old  year_recent  \
465          TU Dresden  Germany      2012        31.9         2016   

     income_recent  improvement  
465           99.7         67.8

> L'università tedesca *TU Dresden* ha avuto il più alto incremento di introiti registrato dal 2012 al 2016.

# 4. For each ranking, consider only the most recent data point. For each university, compute the maximum difference between the rankings (e.g. for *Aarhus University* the value is 122-73=49). Notice that some rankings are expressed as a range

Alcune posizioni vengono espresse sottoforma di intervallo (es. 101-150) nelle classfiche di *shanghai* e *Times*; quest'ultima presenta inoltre il carattere "=" per indicare un pari merito. Per questo è stata creata una funzione che permette di indivuare la parte numerica e riportare la media tra gi estremi nel caso in cui sia presente un intervallo di posizioni, in modo da avere un valore indicativo su cui poter calcolare la differenza massima tra i 3 ranking.

In [20]:
def range_to_mean (dt, column):
    dt[['inf', 'sup']] = dt[column].str.extract('(\d+)-?(\d+)?')
    dt['inf'] = pd.to_numeric(dt['inf'])
    dt['sup'] = dt['sup'].fillna(dt['inf'])
    dt['sup'] = pd.to_numeric(dt['sup'])
    dt['rank'] = round( (dt['sup'] + dt['inf']) /2)
    del dt['inf']
    del dt['sup']
    return dt

Viene applicata la funzione ai dati più recenti di *Times* e viene verificato che la colonna risultante sia di tipo decimale.

In [21]:
recent_times = range_to_mean(recent_times, 'world_rank')
print(recent_times['rank'].head())
recent_times.head()

2003    226.0
2056    276.0
1908    106.0
2105    326.0
2404    700.0
Name: rank, dtype: float64


world_rank             university_name         country  teaching  \
2003    201-250          Aalborg University         Denmark      25.1   
2056    251-300            Aalto University         Finland      31.1   
1908       =106           Aarhus University         Denmark      36.9   
2105    301-350      Aberystwyth University  United Kingdom      21.6   
2404    601-800  Adam Mickiewicz University          Poland      20.0   

      international  research  citations  income  total_score  num_students  \
2003           71.0      28.4       73.8    43.7          NaN       17422.0   
2056           65.4      32.8       62.1    61.6          NaN       16099.0   
1908           76.8      50.7       79.8    68.3         57.7       23895.0   
2105           72.2      18.9       67.2    31.3          NaN        9252.0   
2404           25.7      11.0       15.3    28.7          NaN       40633.0   

      student_staff_ratio international_students female_male_ratio  year  \
2003                 15.9                    15%           48 : 52  2016   
2056                 24.2                    17%           32 : 68  2016   
1908                 13.6                    14%           54 : 46  2016   
2105                 19.2                    18%           48 : 52  2016   
2404                 15.6                     1%           71 : 29  2016   

                       univ_name   rank  
2003          aalborg university  226.0  
2056            aalto university  276.0  
1908           aarhus university  106.0  
2105      aberystwyth university  326.0  
2404  adam mickiewicz university  700.0

Vengono estratti i dati più recenti di *shanghai*, viene applicata la funzione e verificato che la colonna risultante sia di tipo decimale.

In [22]:
recent_shanghai = shanghai.iloc[shanghai.groupby('univ_name')['year'].idxmax()].copy()
recent_shanghai = range_to_mean(recent_shanghai, 'world_rank')
print(recent_shanghai['rank'].head())
recent_shanghai.head()

4697    350.0
4797    450.0
4469     73.0
4497    126.0
4798    450.0
Name: rank, dtype: float64


world_rank                       university_name national_rank  \
4697    301-400                    Aalborg University             5   
4797    401-500                      Aalto University           4-6   
4469         73                     Aarhus University             2   
4497    101-150              Aix Marseille University           5-6   
4798    401-500  Aristotle University of Thessaloniki             2   

      total_score  alumni  award  hici    ns   pub   pcp  year  \
4697          NaN     0.0    0.0  11.2   4.6  30.4  16.8  2015   
4797          NaN     0.0    0.0   0.0  10.4  33.4  17.0  2015   
4469         27.3    11.5   22.1  12.3  25.8  51.8  31.0  2015   
4497          NaN    13.6    0.0  15.2  20.3  50.1  27.6  2015   
4798          NaN     0.0    0.0   4.9   1.5  34.0  16.8  2015   

                                 univ_name   rank  
4697                    aalborg university  350.0  
4797                      aalto university  450.0  
4469                     aarhus university   73.0  
4497              aix marseille university  126.0  
4798  aristotle university of thessaloniki  450.0

Per i dati più recenti di *Cwur* non è necessario applicare la funzione perchè non presenta problemi, infatti il tipo di dati viene letto come numero intero.

In [23]:
recent_cwur = cwur.iloc[cwur.groupby('univ_name')['year'].idxmax()].copy()
print(recent_cwur['world_rank'].head())
recent_cwur.head()

1764    565
1620    421
1321    122
2013    814
2089    890
Name: world_rank, dtype: int64


world_rank                       university_name         country  \
1764         565                    Aalborg University         Denmark   
1620         421                      Aalto University         Finland   
1321         122                     Aarhus University         Denmark   
2013         814                Aberystwyth University  United Kingdom   
2089         890  Adam Mickiewicz University in Poznań          Poland   

      national_rank  quality_of_education  alumni_employment  \
1764              5                   367                439   
1620              4                   367                224   
1321              2                   310                165   
2013             59                   367                567   
2089              5                   150                567   

      quality_of_faculty  publications  influence  citations  broad_impact  \
1764                 218           596        660        287         520.0   
1620                 218           336        403        645         424.0   
1321                 106            91         81        101          95.0   
2013                 218           842        552        511         754.0   
2089                 218           675        775        812         932.0   

      patents  score  year                             univ_name  
1764      401  44.59  2015                    aalborg university  
1620      246  45.10  2015                      aalto university  
1321      154  49.65  2015                     aarhus university  
2013      607  44.22  2015                aberystwyth university  
2089      657  44.14  2015  adam mickiewicz university in poznań

## Controllo della similarità tra università appartenenti a ranking diversi

In [4]:
from two_lists_similarity import Calculate_Similarity as cs

Inizialmente si fa un confronto dei nomi delle università tra quelle di *Shanghai* e *Times*.

Viene creato un oggetto ai fini del calcolo delle similarità e successivamente vengono svolti 3 passaggi:
* si confronta ogni elemento nell'elenco di input con tutti gli elementi nell'elenco di riferimento
* si calcolano i punteggi di somiglianza per ciascuno dei confronti
* l'output mostra l'elemento dell'elenco di input con il corrispondente elemento dell'elenco di riferimento che ha il punteggio di somiglianza più alto

In [ ]:
shangai_to_times_Obj = cs(recent_shanghai['univ_name'], recent_times['univ_name'])
shangai_to_times_Obj.fuzzy_match_output(output_csv_name = 'shanghai_to_times_similarities.csv', output_csv_path = r'../similarities')

Il dataframe di output viene salvato in un file csv per avere traccia delle similarità ottenute.

In [26]:
s_t_sim = pd.read_csv('C:/Users/marco/Desktop/Uni/Computer Science (FCS)/progetto/shanghai_to_times_similarities.csv')
s_t_sim

input_list_item  \
0                      aalborg university   
1                        aalto university   
2                       aarhus university   
3                aix marseille university   
4    aristotle university of thessaloniki   
..                                    ...   
646                  yamaguchi university   
647                    yeshiva university   
648                     yonsei university   
649                       york university   
650                   zhejiang university   

                        similar_ref_list_item  similarity_score  
0                          aalborg university              1.00  
1                            aalto university              1.00  
2                           aarhus university              1.00  
3                    aix marseille university              1.00  
4        aristotle university of thessaloniki              1.00  
..                                        ...               ...  
646  agh university of science and technology              0.86  
647                        yeshiva university              1.00  
648                         yonsei university              1.00  
649                           york university              1.00  
650                       zhejiang university              1.00  

[651 rows x 3 columns]

Si effettua un controllo sulle similarità pari a 1 per verificare che ci sia perfetta uguaglianza.

In [27]:
import numpy as np

In [28]:
s_t_sim['Check'] = np.where(s_t_sim['input_list_item'] == s_t_sim['similar_ref_list_item'], 1, 0)
s_t_sim[(s_t_sim['similarity_score'] == 1) & (s_t_sim['Check'] == 0)]

Empty DataFrame
Columns: [input_list_item, similar_ref_list_item, similarity_score, Check]
Index: []

A partire dal valore 0.99 verso il basso, sono stati visualizzati tutti gli abbinamenti per valutare una soglia ottimale. <br>
Allo score di similarità pari a 0.92 si possono già notare alcuni abbinamenti che non riportano alla stessa università.

In [29]:
s_t_sim[s_t_sim['similarity_score'] == 0.92]

input_list_item              similar_ref_list_item  \
70    ecole normale superieure lyon   école normale supérieure de lyon   
464              university of jena                university of genoa   
472             university of koeln                 university of kent   
554  university of roma tor vergata  university of rome ii tor vergata   
630        vrije university brussel         vrije universiteit brussel   

     similarity_score  Check  
70               0.92      0  
464              0.92      0  
472              0.92      0  
554              0.92      0  
630              0.92      0

A livello 0.92 sono presenti degli accoppiamenti di università differenti (es. university of jena - university of genoa). <br>
Di conseguenza la scelta delle università da uniformare sono quelle con un valore compreso tra 0.93 e 0.99, cioè 108.

In [30]:
s_t_change = s_t_sim[(s_t_sim['similarity_score'] != 1) & (s_t_sim['similarity_score'] > 0.92)]
s_t_change.head()

input_list_item  \
6           arizona state university tempe   
24                       brunel university   
28  capital university of medical sciences   
33            catholic university of chile   
44                china medical university   

                      similar_ref_list_item  similarity_score  Check  
6                  arizona state university              0.95      0  
24                 brunel university london              0.95      0  
28               capital medical university              0.95      0  
33  pontifical catholic university of chile              0.95      0  
44          china medical university taiwan              0.95      0

Nella classifica di *Shanghai* viene creata una nuova colonna `key` a partire da quella dei nomi delle università:
- si mantiene lo stesso nome, se Similarità = 1 (*Shanghai* e *Times* hanno stesso nome per la stessa università)
- si mantiene lo stesso nome, se Similarità <= 0.92 (in *Times* non è presente una università con un nome simile)
- si cambia il nome con quello riportato in *Times* per i restanti casi (stessa università ma con nome scritto in maniera diversa)

In [32]:
recent_shanghai = pd.merge(recent_shanghai, s_t_change[['input_list_item', 'similar_ref_list_item']], left_on = 'univ_name', right_on = 'input_list_item',  how = 'left')
recent_shanghai = recent_shanghai.rename(columns = {'similar_ref_list_item':'key'})
recent_shanghai['key'] = recent_shanghai['key'].fillna(recent_shanghai['univ_name'])
recent_shanghai.drop('input_list_item', axis = 1, inplace = True)
recent_shanghai.head()

world_rank                       university_name national_rank  total_score  \
0    301-400                    Aalborg University             5          NaN   
1    401-500                      Aalto University           4-6          NaN   
2         73                     Aarhus University             2         27.3   
3    101-150              Aix Marseille University           5-6          NaN   
4    401-500  Aristotle University of Thessaloniki             2          NaN   

   alumni  award  hici    ns   pub   pcp  year  \
0     0.0    0.0  11.2   4.6  30.4  16.8  2015   
1     0.0    0.0   0.0  10.4  33.4  17.0  2015   
2    11.5   22.1  12.3  25.8  51.8  31.0  2015   
3    13.6    0.0  15.2  20.3  50.1  27.6  2015   
4     0.0    0.0   4.9   1.5  34.0  16.8  2015   

                              univ_name   rank  \
0                    aalborg university  350.0   
1                      aalto university  450.0   
2                     aarhus university   73.0   
3              aix marseille university  126.0   
4  aristotle university of thessaloniki  450.0   

                                    key  
0                    aalborg university  
1                      aalto university  
2                     aarhus university  
3              aix marseille university  
4  aristotle university of thessaloniki

Si crea una lista contente tutte i nomi delle università di *Times* + tutti i nomi delle università di *Shanghai* che non hanno avuto un match di similarità con *Times*. La lista ottenuta conterrà tutti i possibili casi di matching delle prime due tabelle.

In [34]:
shanghai_times = list(recent_times['univ_name']) + list(s_t_sim[s_t_sim['similarity_score'] <= 0.92]['input_list_item'])

Si confrontato tutti i nomi di *Cwur* con tutti i nomi della lista creata.

In [ ]:
cwur_to_times_and_shanghai_Obj = cs(recent_cwur['univ_name'], shanghai_times)
cwur_to_times_and_shanghai_Obj.fuzzy_match_output(output_csv_name = 'cwur_to_times_and_shanghai_similarities.csv', output_csv_path = r'../similarities')

In [36]:
c_ts_sim = pd.read_csv('C:/Users/marco/Desktop/Uni/Computer Science (FCS)/progetto/cwur_to_times_and_shanghai_similarities.csv')
c_ts_sim['Check'] = np.where(c_ts_sim['input_list_item'] == c_ts_sim['similar_ref_list_item'], 1, 0)
c_ts_sim[(c_ts_sim['similarity_score'] == 1) & (c_ts_sim['Check'] == 0)]

Empty DataFrame
Columns: [input_list_item, similar_ref_list_item, similarity_score, Check]
Index: []

Alla valore 0.94 sono presenti accoppiamenti di 2 università differenti, quindi si opta per una soglia di 0.95.

In [37]:
c_ts_sim[c_ts_sim['similarity_score'] == 0.94]

input_list_item  \
342  nanjing university of aeronautics and astronau...   
417              paul sabatier university toulouse iii   
548                                    toho university   

                                 similar_ref_list_item  similarity_score  \
342  beijing university of aeronautics and astronau...              0.94   
417                paul sabatier university toulouse 3              0.94   
548                                  tohoku university              0.94   

     Check  
342      0  
417      0  
548      0

Nella classifica di *Cwur* viene creata una nuova colonna `key` a partire da quella dei nomi delle università che avrà:
- stesso nome, se Similarità = 1 (*Cwur* ha lo stesso nome presente in *Times* o *Shangai*)
- stesso nome, se Similarità <= 0.94 (in *Times* e *Shangai* non è presente una università con un nome simile)
- si cambia il nome con quello riportato nella lista per i restanti casi, quindi verrà assegnato il nome di *Times* oppure quello di *Shanghai* (se il nome non è simile a nessuno di *Times*)

In [38]:
c_ts_change = c_ts_sim[(c_ts_sim['similarity_score'] != 1) & (c_ts_sim['similarity_score'] > 0.94)]
recent_cwur = pd.merge(recent_cwur, c_ts_change[['input_list_item', 'similar_ref_list_item']], left_on = 'univ_name', right_on = 'input_list_item',  how = 'left')
recent_cwur = recent_cwur.rename(columns = {'similar_ref_list_item':'key'})
recent_cwur['key'] = recent_cwur['key'].fillna(recent_cwur['univ_name'])
recent_cwur.drop('input_list_item', axis = 1, inplace = True)
recent_cwur.head()

world_rank                       university_name         country  \
0         565                    Aalborg University         Denmark   
1         421                      Aalto University         Finland   
2         122                     Aarhus University         Denmark   
3         814                Aberystwyth University  United Kingdom   
4         890  Adam Mickiewicz University in Poznań          Poland   

   national_rank  quality_of_education  alumni_employment  quality_of_faculty  \
0              5                   367                439                 218   
1              4                   367                224                 218   
2              2                   310                165                 106   
3             59                   367                567                 218   
4              5                   150                567                 218   

   publications  influence  citations  broad_impact  patents  score  year  \
0           596        660        287         520.0      401  44.59  2015   
1           336        403        645         424.0      246  45.10  2015   
2            91         81        101          95.0      154  49.65  2015   
3           842        552        511         754.0      607  44.22  2015   
4           675        775        812         932.0      657  44.14  2015   

                              univ_name                         key  
0                    aalborg university          aalborg university  
1                      aalto university            aalto university  
2                     aarhus university           aarhus university  
3                aberystwyth university      aberystwyth university  
4  adam mickiewicz university in poznań  adam mickiewicz university

Per comodità nell'unione, anche in *Times* viene creata una colonna chiamata `key`.

In [39]:
recent_times['key'] = recent_times['univ_name']

Struttura finale delle similarità in cui un nome simile ($\approx$) tra due classifiche che fanno riferimento alla stessa università, viene sostituito ($\Longrightarrow$) con quello di una delle due:
* *Shanghai* $\approx$ *Times* $\Longrightarrow$ *Times*
* *Cwur* $\approx$ *Times* $\Longrightarrow$ *Times*
* *Cwur* $\approx$ *Shanghai* $\Longrightarrow$ *Shanghai*
* *Cwur* $\approx$ *Shanghai* $\approx$ *Times* $\Longrightarrow$ *Times*

## Unione dei ranking

La tabella risultante deve avere tutte le università, di conseguenza l'unione delle tabelle deve restituire un numero di righe $\geq$ alla tabella con più righe.

In [40]:
print(recent_times.shape)
print(recent_shanghai.shape)
print(recent_cwur.shape)

(818, 17)
(651, 14)
(1024, 16)


Si uniscono le tabelle dei 3 ranking sulla chiave `key` utilizzando la modalità *outer*, cioè inserendo nella tabella finale sia le università in comune, sia quelle non in comune. <br>
Si verifica inoltre che il numero di righe è 1291 $\geq$ 1024.

In [41]:
fuso_temp = pd.merge(recent_times, recent_shanghai, on = 'key', how = 'outer', suffixes = ('_times', '_shanghai'))
fuso = pd.merge(fuso_temp, recent_cwur, on = 'key', how = 'outer', suffixes = ('', '_cwur'))
fuso = fuso.rename(columns = {'world_rank':'rank_cwur'})
print(fuso.shape)
fuso[['university_name_times', 'rank_times', 'rank_shanghai', 'rank_cwur']].head()

(1291, 45)


university_name_times  rank_times  rank_shanghai  rank_cwur
0          Aalborg University       226.0          350.0      565.0
1            Aalto University       276.0          450.0      421.0
2           Aarhus University       106.0           73.0      122.0
3      Aberystwyth University       326.0            NaN      814.0
4  Adam Mickiewicz University       700.0            NaN      890.0

Una volta ottenute le posizioni in classifiche dei dati più recenti per ogni università, è stata creata una funzione in grado di calcolare la differenza massima tra questi, ovviando al problema dei valori mancanti creati creati dall'unione delle tabelle. <br>
Nel caso in cui 2 dei 3 ranking siano valori mancanti, viene restituito *NaN*

In [42]:
def diff_rank (a, b, c):
    diff = float()
    if (pd.notna(a) & pd.notna(b) & pd.notna(c)):
        diff = max(a, b, c) - min(a, b, c)
    elif (pd.notna(a) & pd.notna(b)):
            diff = max(a, b) - min(a, b)
    elif (pd.notna(a) & pd.notna(c)):
            diff = max(a, c) - min(a, c)
    elif (pd.notna(b) & pd.notna(c)):
            diff = max(b, c) - min(b, c)
    else:
        diff = float('NaN')
    return diff

Viene applicata la funzione creata precedentemente riga per riga, in modo da creare una nuova colonna corrispondente alla differenza più grande tra le 3 posizioni.

In [43]:
fuso['diff_max'] = fuso.apply(lambda row: diff_rank(row['rank_times'], row['rank_shanghai'], row['rank_cwur']), axis = 1)
fuso.sort_values(by = 'diff_max', inplace = True, ascending = False)
fuso[['university_name_times', 'country', 'rank_times', 'rank_shanghai', 'rank_cwur', 'diff_max']].dropna()

university_name_times                   country  \
879                    Zhejiang University                     China   
780                   University of Sydney                 Australia   
219              King Abdulaziz University              Saudi Arabia   
681            University of Massachusetts  United States of America   
680            University of Massachusetts  United States of America   
..                                     ...                       ...   
577                  University of Chicago  United States of America   
265  Massachusetts Institute of Technology  United States of America   
568                University of Cambridge            United Kingdom   
870                        Yale University  United States of America   
443                    Stanford University  United States of America   

     rank_times  rank_shanghai  rank_cwur  diff_max  
879       276.0          126.0      991.0     865.0  
780        56.0          126.0      882.0     826.0  
219       276.0          176.0      995.0     819.0  
681       141.0          126.0      884.0     758.0  
680       141.0          126.0      825.0     699.0  
..          ...            ...        ...       ...  
577        10.0            9.0        8.0       2.0  
265         5.0            3.0        3.0       2.0  
568         4.0            5.0        4.0       1.0  
870        12.0           11.0       11.0       1.0  
443         3.0            2.0        2.0       1.0  

[517 rows x 6 columns]

> L'università *Zhejiang University* dell'Arabia Saudita prensenta la differenza massima più elevata tra i ranking *shanghai* e *Cwur*, mentre 2 università americane e una inglese presentano differenze massime di una sola posizione tra le classifiche.

# 5. Consider only the most recent data point of the *times* dataset. Compute the number of male and female students for each country.

Attraverso l'utilizzo di un'espressione regolare si ricava la proporzione di studenti maschi e femmine per ciascuna università e si converte in formato numerico.

In [44]:
recent_times[['female_ratio', 'male_ratio']] = recent_times['female_male_ratio'].str.extract('(\d+) : (\d+)')
recent_times['female_ratio'] = pd.to_numeric(recent_times['female_ratio'])
recent_times['male_ratio'] = pd.to_numeric(recent_times['male_ratio'])
recent_times.head()

world_rank             university_name         country  teaching  \
2003    201-250          Aalborg University         Denmark      25.1   
2056    251-300            Aalto University         Finland      31.1   
1908       =106           Aarhus University         Denmark      36.9   
2105    301-350      Aberystwyth University  United Kingdom      21.6   
2404    601-800  Adam Mickiewicz University          Poland      20.0   

      international  research  citations  income  total_score  num_students  \
2003           71.0      28.4       73.8    43.7          NaN       17422.0   
2056           65.4      32.8       62.1    61.6          NaN       16099.0   
1908           76.8      50.7       79.8    68.3         57.7       23895.0   
2105           72.2      18.9       67.2    31.3          NaN        9252.0   
2404           25.7      11.0       15.3    28.7          NaN       40633.0   

      student_staff_ratio international_students female_male_ratio  year  \
2003                 15.9                    15%           48 : 52  2016   
2056                 24.2                    17%           32 : 68  2016   
1908                 13.6                    14%           54 : 46  2016   
2105                 19.2                    18%           48 : 52  2016   
2404                 15.6                     1%           71 : 29  2016   

                       univ_name   rank                         key  \
2003          aalborg university  226.0          aalborg university   
2056            aalto university  276.0            aalto university   
1908           aarhus university  106.0           aarhus university   
2105      aberystwyth university  326.0      aberystwyth university   
2404  adam mickiewicz university  700.0  adam mickiewicz university   

      female_ratio  male_ratio  
2003          48.0        52.0  
2056          32.0        68.0  
1908          54.0        46.0  
2105          48.0        52.0  
2404          71.0        29.0

Dalla proporzione e dal totale degli studenti, si calcola il numero effettivo di studentesse per ogni università. <br>
I restanti saranno studenti maschi. 

In [45]:
recent_times['female'] = round(recent_times['num_students'] * recent_times['female_ratio'] / 100)
recent_times['male'] = recent_times['num_students'] - recent_times['female']
recent_times[['university_name', 'country', 'female', 'male']].head()

university_name         country   female     male
2003          Aalborg University         Denmark   8363.0   9059.0
2056            Aalto University         Finland   5152.0  10947.0
1908           Aarhus University         Denmark  12903.0  10992.0
2105      Aberystwyth University  United Kingdom   4441.0   4811.0
2404  Adam Mickiewicz University          Poland  28849.0  11784.0

Per ottenere il numero totale di studenti e studentesse per ciascuna nazione, si aggregano i valori per `country` e si fa una somma per entrambi i generi.

In [46]:
recent_times.groupby('country')[['female', 'male']].sum()

female       male
country                                       
Argentina                   67191.0    41182.0
Australia                  391736.0   321640.0
Austria                     68364.0    66113.0
Bangladesh                  21323.0    41393.0
Belarus                     20219.0     9084.0
...                             ...        ...
Uganda                      18670.0    18670.0
Ukraine                     17846.0    19250.0
United Arab Emirates         9516.0     4931.0
United Kingdom             711814.0   613028.0
United States of America  1601453.0  1556322.0

[70 rows x 2 columns]

# 6. Find the universities where the ratio between female and male is below the average ratio (computed over all universities)

Calcolando il rapporto tra studentesse e studenti per ogni università, viene rilevato un valore "anomalo": infatti nell'università `Ewha Womans University` sono presenti esclusivamente studentesse, quindi avendo 0 al denominatore, il rapporto è pari a infinito.

In [47]:
recent_times['female/male'] = recent_times['female'] / recent_times['male']
recent_times[recent_times['university_name']=='Ewha Womans University'] # only females in this university

world_rank         university_name      country  teaching  international  \
2227    401-500  Ewha Womans University  South Korea      23.9           34.3   

      research  citations  income  total_score  num_students  ...  \
2227      18.3       50.4    57.6          NaN       17625.0  ...   

      female_male_ratio  year               univ_name   rank  \
2227            100 : 0  2016  ewha womans university  450.0   

                         key  female_ratio male_ratio   female  male  \
2227  ewha womans university         100.0        0.0  17625.0   0.0   

      female/male  
2227          inf  

[1 rows x 22 columns]

Per quanto descritto in precedenza, si scarta il valore corrispondente all'università di sole femmine in modo da poter fare una media tra i valori non nulli di `female/male` e ottenere il rapporto $\frac{\textit{Femmine}}{\textit{Maschi}}$ medio.

In [48]:
avg_ratio = recent_times['female/male'].drop(labels = 2227).mean()
avg_ratio

1.0811006226646445

A questo punto vengono riportate solo le università che presentano il rapporto $\frac{\textit{Femmine}}{\textit{Maschi}}$ <  1.0811006226646445

In [49]:
times_ratio = recent_times[recent_times['female/male'] < avg_ratio]
times_ratio[['university_name', 'female/male']]

university_name  female/male
2003                        Aalborg University     0.923170
2056                          Aalto University     0.470631
2105                    Aberystwyth University     0.923093
2406                           Ajou University     0.492541
2408                     Alexandria University     0.851845
...                                        ...          ...
2104                       Zhejiang University     0.694898
1856                  École Normale Supérieure     0.851852
2013          École Normale Supérieure de Lyon     0.961096
1904                       École Polytechnique     0.219378
1833  École Polytechnique Fédérale de Lausanne     0.369898

[346 rows x 2 columns]

> Il rapporto medio ottenuto indica che in media le università hanno una maggioranza di studenti di genere femminile rispetto a quello maschile. <br>
Sono presenti 346 università che hanno un rapporto inferiore alla media.

# 7. For each country, compute the fraction of the students in the country that are in one of the universities computed in the previous point (that is, the denominator of the ratio is the total number of students over all universities in the country).

Per calcolare il numeratore dell'assegnamento, vengono considerate solo le università che hanno un rapporto $\frac{\textit{Femmine}}{\textit{Maschi}}$ minore della media, individuate nella tabella ottenuta al punto precedente dai dati più recenti di *Times*. <br>
Viene calcolata la somma di tutti gli studenti raggruppati per `country`.

In [50]:
stud_ratio = times_ratio.groupby('country', as_index = False)['num_students'].sum()
stud_ratio.rename(columns = {'num_students':'country_students_ratio'}, inplace = True)
stud_ratio.head()

country  country_students_ratio
0   Australia                146235.0
1     Austria                 61033.0
2  Bangladesh                 62716.0
3      Brazil                278774.0
4      Canada                 86779.0

Per calcolare il numeratore dell'assegnamento, vengono considerate tutte le università riferite ai dati più recenti di *Times*. <br>
Viene calcolata la somma di tutti gli studenti raggruppati per `country`. <br>
Le somme pari a 0 vengono considerate come *NaN*, in modo da non scaturire un errore causato dal denominatore di una frazione nullo. Queste molto probabilmente corrispondono ad errori di inserimento dati o comunque a valori mancanti.

In [51]:
all_stud_for_country = recent_times.groupby('country', as_index = False)['num_students'].sum()
all_stud_for_country.rename(columns = {'num_students':'all_country_students'}, inplace = True)
all_stud_for_country.replace(0, float('NaN'), inplace = True)
all_stud_for_country

country  all_country_students
0                  Argentina              108373.0
1                  Australia              743627.0
2                    Austria              134477.0
3                 Bangladesh               62716.0
4                    Belarus               29303.0
..                       ...                   ...
65                    Uganda               37340.0
66                   Ukraine               37096.0
67      United Arab Emirates               14447.0
68            United Kingdom             1324842.0
69  United States of America             3280303.0

[70 rows x 2 columns]

Viene eseguita un'unione con chiave `country` di tipo "right" tra la prima tabella ottenuta e la seconda ripulita dai valori mancanti, per assicurarsi che vengano riportate tutte le nazioni con almeno uno studente di *Times*. Dopoidchè viene sostituito il valore 0 laddove venga generato un *NaN*, in modo che venga comunque calcolata una percentuale, seppur pari a 0.

In [52]:
students_country = pd.merge(stud_ratio, all_stud_for_country.dropna(), on = 'country', how = 'right')
students_country['country_students_ratio'] = students_country['country_students_ratio'].fillna(0)
students_country['perc'] = round(students_country['country_students_ratio'] / students_country['all_country_students'] * 100, 2)
students_country[['country', 'country_students_ratio', 'all_country_students', 'perc']].sort_values(by = 'perc', ascending = False)

country  country_students_ratio  all_country_students   perc
35          Oman                  7089.0                7089.0  100.0
30       Morocco                 56060.0               56060.0  100.0
50        Uganda                 37340.0               37340.0  100.0
15         Ghana                 37274.0               37274.0  100.0
25         Kenya                 72676.0               72676.0  100.0
..           ...                     ...                   ...    ...
59       Iceland                     0.0               13960.0    0.0
57       Belgium                     0.0              169661.0    0.0
56       Belarus                     0.0               29303.0    0.0
55     Argentina                     0.0              108373.0    0.0
69  South Africa                     0.0              330198.0    0.0

[70 rows x 4 columns]

> 4 delle prime 5 righe che vengono visualizzate, sono Paesi africani che hanno in ogni università un rapporto $\frac{\textit{Femmine}}{\textit{Maschi}}$ inferiore alla media. Mentre tutte le università presenti nelle nazioni con una percentuale pari a 0 presentano un numero superiore di studenti di genere femminile rispetto al genere maschile, quindi un rapporto superiore a circa 1,08.

# 8. Read the file `educational_attainment_supplementary_data.csv`, discarding any row with missing `country_name` or `series_name`

In [53]:
attainment.dropna(subset = ('country_name', 'series_name'), inplace = True)
attainment

country_name                                        series_name  1985  \
0      Afghanistan  Barro-Lee: Average years of primary schooling,...  0.33   
1      Afghanistan  Barro-Lee: Average years of primary schooling,...  1.03   
2      Afghanistan  Barro-Lee: Average years of primary schooling,...  0.83   
3      Afghanistan  Barro-Lee: Average years of primary schooling,...  2.34   
4      Afghanistan  Barro-Lee: Average years of primary schooling,...  0.54   
...            ...                                                ...   ...   
79045     Zimbabwe  UIS: Percentage of population age 25+ with som...   NaN   
79046     Zimbabwe  UIS: Percentage of population age 25+ with som...   NaN   
79047     Zimbabwe  UIS: Percentage of population age 25+ with unk...   NaN   
79048     Zimbabwe  UIS: Percentage of population age 25+ with unk...   NaN   
79049     Zimbabwe  UIS: Percentage of population age 25+ with unk...   NaN   

       1986  1987  1990  1991  1992  1993  1995  ...  2005  2006  2007  2008  \
0       NaN   NaN  0.44   NaN   NaN   NaN  0.57  ...  0.86   NaN   NaN   NaN   
1       NaN   NaN  1.26   NaN   NaN   NaN  1.54  ...  2.18   NaN   NaN   NaN   
2       NaN   NaN  0.95   NaN   NaN   NaN  1.26  ...  1.01   NaN   NaN   NaN   
3       NaN   NaN  2.22   NaN   NaN   NaN  2.37  ...  2.26   NaN   NaN   NaN   
4       NaN   NaN  0.92   NaN   NaN   NaN  0.94  ...  2.00   NaN   NaN   NaN   
...     ...   ...   ...   ...   ...   ...   ...  ...   ...   ...   ...   ...   
79045   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   NaN   NaN   NaN   
79046   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   NaN   NaN   NaN   
79047   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   NaN   NaN   NaN   
79048   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   NaN   NaN   NaN   
79049   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   NaN   NaN   NaN   

       2009  2010  2011      2012  2013  2015  
0       NaN  1.27   NaN       NaN   NaN   NaN  
1       NaN  2.64   NaN       NaN   NaN   NaN  
2       NaN  2.45   NaN       NaN   NaN   NaN  
3       NaN  3.55   NaN       NaN   NaN   NaN  
4       NaN  1.29   NaN       NaN   NaN   NaN  
...     ...   ...   ...       ...   ...   ...  
79045   NaN   NaN   NaN  13.84323   NaN   NaN  
79046   NaN   NaN   NaN  16.68491   NaN   NaN  
79047   NaN   NaN   NaN  11.99412   NaN   NaN  
79048   NaN   NaN   NaN   5.77150   NaN   NaN  
79049   NaN   NaN   NaN   9.14588   NaN   NaN  

[79050 rows x 29 columns]

> La tabella risultante è stata ripulita da tutte quelle righe che presentano valori mancanti in `country_name` e in `series_name`.

# 9. From `attainment` build a dataframe with the same data, but with 4 columns: `country_name`, `series_name`, `year`, `value`

Per soddisfare la richiesta viene utilizzata la funzione di pandas *melt*, la quale permette la trasformazione della tabella caricata in precedenza da *wide* a *long*. In questo modo la coppia `country_name`-`series_name` viene replicata per ciascun anno con il corrispettivo valore.

In [54]:
att_long = pd.melt(attainment, id_vars=['country_name', 'series_name'], var_name='year').dropna()
att_long

country_name  \
0               Afghanistan   
1               Afghanistan   
2               Afghanistan   
3               Afghanistan   
4               Afghanistan   
...                     ...   
2054020  West Bank and Gaza   
2054021  West Bank and Gaza   
2054022  West Bank and Gaza   
2054023  West Bank and Gaza   
2054024  West Bank and Gaza   

                                               series_name  year    value  
0        Barro-Lee: Average years of primary schooling,...  1985  0.33000  
1        Barro-Lee: Average years of primary schooling,...  1985  1.03000  
2        Barro-Lee: Average years of primary schooling,...  1985  0.83000  
3        Barro-Lee: Average years of primary schooling,...  1985  2.34000  
4        Barro-Lee: Average years of primary schooling,...  1985  0.54000  
...                                                    ...   ...      ...  
2054020  UIS: Percentage of population age 25+ with som...  2013  1.48356  
2054021  UIS: Percentage of population age 25+ with som...  2013  1.88820  
2054022  UIS: Percentage of population age 25+ with unk...  2013  0.00000  
2054023  UIS: Percentage of population age 25+ with unk...  2013  0.00000  
2054024  UIS: Percentage of population age 25+ with unk...  2013  0.00000  

[318290 rows x 4 columns]

# 10. For each university, find the number of rankings in which they appear (it suffices to appear in one year for each ranking).

Per ottenere una lista contenente tutte le università di tutti i ranking viene sfruttata la colonna `key` della tabella creata al punto **4.** tramite l'unione di tipo *outer* dei dei ranking con i dati più recenti. Dopodichè viene verificato per ognuna di esse in quanti e quali dei 3 ranking compaiono. 

In [55]:
def appearence (univ):
    lista = []
    if univ in list(recent_times['key']):
        lista.append('Times')
    if univ in list(recent_shanghai['key']):
        lista.append('shanghai')
    if univ in list(recent_cwur['key']):
        lista.append('Cwur')
    return lista

In [56]:
appearences = fuso['key'].to_frame()
appearences['rankings'] = appearences.apply(lambda row: appearence(row['key']), axis = 1)
appearences['appearences'] = appearences.apply(lambda row: len(row['rankings']), axis = 1)
appearences

key                 rankings  appearences
879               zhejiang university  [Times, shanghai, Cwur]            3
780              university of sydney  [Times, shanghai, Cwur]            3
219         king abdulaziz university  [Times, shanghai, Cwur]            3
671            university of mannheim            [Times, Cwur]            2
681       university of massachusetts  [Times, shanghai, Cwur]            3
...                               ...                      ...          ...
1286             zhengzhou university                   [Cwur]            1
1287           åbo akademi university                   [Cwur]            1
1288           école centrale de lyon                   [Cwur]            1
1289             école centrale paris                   [Cwur]            1
1290  école polytechnique de montréal                   [Cwur]            1

[1291 rows x 3 columns]

# 11. In the times ranking, compute the number of times each university appears

Ai fini dell'assegnamento, si effettua un raggruppamento per università e si conta il numero delle righe raggruppate per ognuna.

In [57]:
univ_appearences = times.groupby('univ_name').size()
univ_appearences

univ_name
aalborg university                          5
aalto university                            5
aarhus university                           6
aberystwyth university                      5
adam mickiewicz university                  1
                                           ..
école normale supérieure                    6
école normale supérieure de lyon            6
école polytechnique                         6
école polytechnique fédérale de lausanne    6
örebro university                           1
Length: 818, dtype: int64

# 12. Find the universities that appear at most twice in the times ranking.

Utilizzando quanto elaborato al punto precedente, vengono riportato solo le università che appaiono al massimo 2 volte nel ranking *Times*.

In [58]:
univ_appearences[univ_appearences <= 2]

univ_name
adam mickiewicz university                  1
agh university of science and technology    1
aix marseille university                    1
ajou university                             1
alexandru ioan cuza university              1
                                           ..
yokohama city university                    1
yokohama national university                1
yuan ze university                          2
yıldız technical university                 1
örebro university                           1
Length: 408, dtype: int64

# 13. The universities that, in any year, have the same position in all three rankings (they must have the same position in a year).

Ad ogni università delle classifiche di partenza viene associato il valore corrispondente della chiave creata al punto 4 tramite una *left join*.

In [59]:
times['key'] = times['univ_name']

shanghai = pd.merge(shanghai, s_t_change[['input_list_item', 'similar_ref_list_item']], left_on = 'univ_name', right_on = 'input_list_item',  how = 'left')
shanghai = shanghai.rename(columns = {'similar_ref_list_item':'key'})
shanghai['key'] = shanghai['key'].fillna(shanghai['univ_name'])

cwur = pd.merge(cwur, c_ts_change[['input_list_item', 'similar_ref_list_item']], left_on = 'univ_name', right_on = 'input_list_item',  how = 'left')
cwur = cwur.rename(columns = {'similar_ref_list_item':'key'})
cwur['key'] = cwur['key'].fillna(cwur['univ_name'])

Per poter individuare le università che hanno la stessa posizione in classifica nello stesso anno, si è optato per una unione di tipo *inner*. Per fare ciò è necessario che le chiavi dell'unione abbiamo stesso formato, quindi il ranking di *Cwur* viene trasformato da tipo numerico a tipo stringa. Mentre viene rimosso il carattere "=" in modo che tutte le posizioni di pari merito all'interno di *Times* abbiano solo la parte numerica.

In [60]:
cwur['world_rank'] = cwur['world_rank'].apply(str)
times['world_rank'] = times['world_rank'].str.replace('=', '')

A questo punto si può procedere con l'unione dei ranking con chiave multipla formata da `key`, `year` e `world_rank`, il risultato sarà formato da quelle righe che hanno i valori delle 3 colonne esattamente identici.

In [61]:
fusion_temp = pd.merge(cwur, times, on = ['key', 'year', 'world_rank'], suffixes = ('_cwur', '_times'))
punctual_fusion = pd.merge(fusion_temp, shanghai, on = ['key', 'year', 'world_rank'], suffixes = ('', '_shanghai'))
punctual_fusion[['university_name', 'country_times', 'year', 'world_rank']]

university_name             country_times  year world_rank
0  Stanford University  United States of America  2013          2

> L'unica ad avere la stessa posizione nello stesso anno in tutte le classifiche è l'università americana *Stanford University* che nel 2013 ha avuto la seconda posizione in tutti e 3 i ranking.

Come riportato in vari punti, la classifiche *Times* e *shanghai* riportano alcune posizioni in classifica come intervallo: con il metodo precedente si dovrebbero avere intervalli definiti per tutte le classifiche, cosa che in realtà non avviene. Per tenere in considerazione anche i range, si è deciso di riportare anche quelle università che potrebbero avere stessa posizione in tutte e 3 le classifiche.

Viene effettuata una nuova unione dei ranking con chiave la coppia `key`-`year`, in modo da avere per ognuna i 3 posizionamenti nei diversi ranking.

In [62]:
fusion_temp = pd.merge(cwur, times, on = ['key', 'year'], suffixes = ('', '_t'))
range_fusion = pd.merge(fusion_temp, shanghai, on = ['key', 'year'], suffixes = ('', '_s'))
range_fusion.rename(columns = {'world_rank':'WR_Cwur', 'world_rank_t':'WR_Times', 'world_rank_s':'WR_shanghai'}, inplace = True)
range_fusion[['university_name', 'year', 'WR_Cwur', 'WR_Times', 'WR_shanghai']]

university_name  year WR_Cwur WR_Times  \
0                                   Harvard University  2012       1        2   
1                Massachusetts Institute of Technology  2012       2        7   
2                                  Stanford University  2012       3        2   
3                              University of Cambridge  2012       4        6   
4                   California Institute of Technology  2012       5        1   
..                                                 ...   ...     ...      ...   
781                 Swinburne University of Technology  2015     688  351-400   
782                    Sharif University of Technology  2015     697  301-350   
783  Rutgers, The State University of New Jersey - ...  2015     723      144   
784                    National Sun Yat-sen University  2015     803  351-400   
785                   University of Technology, Sydney  2015     815  226-250   

    WR_shanghai  
0             1  
1             3  
2             2  
3             5  
4             6  
..          ...  
781     301-400  
782     401-500  
783          64  
784     401-500  
785     301-400  

[786 rows x 5 columns]

Si crea una funzione che permette di estrarre l'estremo inferiore e l'estremo superiore degli intervalli. Nel caso di valori puntuali viene riportato lo stesso numero sia come estremo inferiore sia come estremo superiore.

In [63]:
def range_to_inf_and_sup (dt, column, inf, sup):
    dt[[inf, sup]] = dt[column].str.extract('(\d+)-?(\d+)?')
    dt[inf] = pd.to_numeric(dt[inf])
    dt[sup] = dt[sup].fillna(dt[inf])
    dt[sup] = pd.to_numeric(dt[sup])
    return dt

Applicando la funzione alle 2 colonne che presentano valori intervallari, vengono create le colonne contenenti gli estremi di ognuna.

In [64]:
range_fusion = range_to_inf_and_sup(range_fusion, 'WR_Times', 'WRt_inf', 'WRt_sup')
range_fusion = range_to_inf_and_sup(range_fusion, 'WR_shanghai', 'WRs_inf', 'WRs_sup')
range_fusion['WR_Cwur'] = pd.to_numeric(range_fusion['WR_Cwur'])
range_fusion[['university_name', 'year', 'WR_Cwur', 'WR_Times', 'WRt_inf', 'WRt_sup', 'WR_shanghai', 'WRs_inf', 'WRs_sup']]

university_name  year  WR_Cwur  \
0                                   Harvard University  2012        1   
1                Massachusetts Institute of Technology  2012        2   
2                                  Stanford University  2012        3   
3                              University of Cambridge  2012        4   
4                   California Institute of Technology  2012        5   
..                                                 ...   ...      ...   
781                 Swinburne University of Technology  2015      688   
782                    Sharif University of Technology  2015      697   
783  Rutgers, The State University of New Jersey - ...  2015      723   
784                    National Sun Yat-sen University  2015      803   
785                   University of Technology, Sydney  2015      815   

    WR_Times  WRt_inf  WRt_sup WR_shanghai  WRs_inf  WRs_sup  
0          2        2        2           1        1        1  
1          7        7        7           3        3        3  
2          2        2        2           2        2        2  
3          6        6        6           5        5        5  
4          1        1        1           6        6        6  
..       ...      ...      ...         ...      ...      ...  
781  351-400      351      400     301-400      301      400  
782  301-350      301      350     401-500      401      500  
783      144      144      144          64       64       64  
784  351-400      351      400     401-500      401      500  
785  226-250      226      250     301-400      301      400  

[786 rows x 9 columns]

A questo punto, avendo le posizioni della classifica *Cwur* esclusivamente in maniera puntuale, si possono considerare quelle università il cui valore di `WR_Cwur` sia contemporaneamente maggiore degli estremi inferiori e minore di quelli superiori di `WR_Times` e `WR_shanghai`.

In [65]:
range_fusion[(range_fusion['WR_Cwur'] >= range_fusion['WRt_inf']) 
             & (range_fusion['WR_Cwur'] >= range_fusion['WRs_inf'])
             & (range_fusion['WR_Cwur'] <= range_fusion['WRt_sup'])
             & (range_fusion['WR_Cwur'] <= range_fusion['WRs_sup'])][['university_name', 'country', 'year', 'WR_Cwur', 'WR_Times', 'WR_shanghai']]

university_name         country  year  WR_Cwur  \
78                       Stanford University             USA  2013        2   
265                University of Southampton  United Kingdom  2014      146   
310                  Sungkyunkwan University     South Korea  2014      204   
380                     Linköping University          Sweden  2014      302   
392                    University of Vermont             USA  2014      320   
417                        Temple University             USA  2014      390   
643                     Newcastle University  United Kingdom  2015      212   
659                         Laval University          Canada  2015      229   
665       Autonomous University of Barcelona           Spain  2015      241   
685             Medical University of Vienna         Austria  2015      265   
687                   University of Waterloo          Canada  2015      269   
707  University of South Carolina - Columbia             USA  2015      300   
731                     Linköping University          Sweden  2015      355   
740     Washington State University, Pullman             USA  2015      380   
742                      University of Turku         Finland  2015      383   

    WR_Times WR_shanghai  
78         2           2  
265      146     101-150  
310  201-225     201-300  
380  301-350     301-400  
392  301-350     301-400  
417  351-400     301-400  
643  201-225     201-300  
659  226-250     201-300  
665  226-250     201-300  
685  251-275     201-300  
687  251-275     201-300  
707  276-300     201-300  
731  351-400     301-400  
740  351-400     301-400  
742  351-400     301-400

> Oltre all'università americana *Stanford University*, ora compaiono anche quelle università che in uno stesso anno potrebbero potenzialmente avere stessa posizione in classifica ma che non è possibile conoscere a causa dell'aggregazione del dato. L'università inglese *University of Southampton* è una buona candidata a soddisfare l'ipotesi precedente, dato che nel 2014 ha stessa posizione in classifica puntuale in 2 su 3 ranking.